components/calendar/src/calendar_arithmetic.rs

```diff
@@ -34,41 +34,48 @@ impl<C: CalendarArithmetic> ArithmeticDate<C> {
             marker: PhantomData,
         }
     }
+
     #[inline]
-    pub fn offset_date(&mut self, mut offset: DateDuration<C>) {
-        self.year += offset.years;
-        self.month += offset.months as u8;
-        offset.months = 0;
-
-        offset.days += offset.weeks * 7;
-
-        offset.days += self.day as i32 - 1;
-        self.day = 1;
-
-        while offset.days != 0 {
-            if offset.days < 0 {
-                self.month -= 1;
-                let month_days = C::month_days(self.year, self.month);
-                if (-offset.days) > month_days as i32 {
-                    offset.days += month_days as i32;
-                } else {
-                    self.day = 1 + (month_days as i8 + offset.days as i8) as u8;
-                    offset.days = 0;
-                }
+    pub fn offset_days(&mut self, mut day_offset: i32) {
+        while day_offset != 0 {
+            let month_days = C::month_days(self.year, self.month);
+            if self.day as i32 + day_offset > month_days as i32 {
+                self.offset_months(1);
+                day_offset -= month_days as i32;
+            } else if self.day as i32 + day_offset < 1 {
+                self.offset_months(-1);
+                day_offset += C::month_days(self.year, self.month) as i32;
+            } else {
+                self.day = (self.day as i32 + day_offset) as u8;
+                day_offset = 0;
+            }
+        }
+    }
+
+    #[inline]
+    pub fn offset_months(&mut self, mut month_offset: i32) {
+        while month_offset != 0 {
+            let year_months = C::months_for_every_year(self.year);
+            if self.month as i32 + month_offset > year_months as i32 {
+                self.year += 1;
+                month_offset -= year_months as i32;
+            } else if self.month as i32 + month_offset < 1 {
+                self.year -= 1;
+                month_offset += C::months_for_every_year(self.year) as i32;
             } else {
-                let month_days = C::month_days(self.year, self.month);
-
-                if offset.days >= month_days as i32 {
-                    self.month += 1;
-                    offset.days -= month_days as i32;
-                } else {
-                    self.day += offset.days as u8;
-                    offset.days = 0;
-                }
+                self.month = (self.month as i32 + month_offset) as u8;
+                month_offset = 0
             }
         }
     }
 
+    #[inline]
+    pub fn offset_date(&mut self, offset: DateDuration<C>) {
+        self.year += offset.years;
```

Comment:
nit: there will be a similar issue as above when it comes to months that are not valid in the future year. This is hard to handle for now, just drop in a comment that offsetting a year may need some care on lunar calendars

Reply:
Done - added comment.

Although, can we assume that `month = 1` is always valid? If so, I believe we can handle month offset the same way we handle day offset, since `offset_months` already takes `months_for_every_year` as a variable.

e.g.
```
let month_offset = offset.months + self.month as i32 - 1;
self.month = 1;
self.offset_months(month_offset);
```

WDYT?

Reply:
we cannot make that assumption unfortunately

Reply:
(the Chinese lunisolar calendar does have leap months at the beginning of the year)